snakemake/common.rules

```diff
@@ -28,27 +30,43 @@ _READ_LENGTH = 124
 _STAR_GENOME_DIR = os.path.join(
     os.path.dirname(config["reference"]["genome"]), "star-genome-%d" % _READ_LENGTH)
 
+
 # Common functions
 
 def _get_half_cores(_):
-  return max(1, int(config["num_threads"]/2))
+  n = _get_all_cores(_)
+  return max(1, n // 2)
 
 def _get_all_cores(_):
-  return int(config["num_threads"])
+  if "num_threads" in config:
+    return int(config["num_threads"])
+  else:
+    return psutil.cpu_count()
 
-def _get_threads_for_alignment(_):
-  if _mem_gb_for_alignment() < 20:
-    return _get_all_cores(_)
+def _available_memory_gb():
+  if "mem_gb" in config:
+    return int(config["mem_gb"])
   else:
-    return _get_half_cores(_)
+    bytes_per_gb = 1024 ** 3
+    return max(1, psutil.virtual_memory().total // bytes_per_gb)
 
-# 20GB or as close to that as allowed by pipeline resources
-def _mem_gb_for_ram_hungry_jobs():
-  return min(20, int(config["mem_gb"]))
+def _get_threads_for_alignment(_):
+  if _available_memory_gb() < 20:
```

Comment:
actually the old logic was what i was going for: we want to use half the cores in a beefy machine (since if we have 20 CPUs to play with, can afford to use 10 here and run some other job in parallel), or all the cores on a small one (if we have only 2 CPUs, may as well use them all for alignment)